lib/adapter.js

```diff
@@ -47,15 +47,37 @@ adapter.run = async function mochaAdapter (cid, config = { mochaOpts: {} }, spec
     specs.forEach((spec) => mocha.addFile(spec))
     mocha.suite.on('pre-require', () => INTERFACES[config.mochaOpts.ui].forEach(runInFiberContext.bind(null, INTERFACES, config.mochaOpts.ui)))
 
-    await runHook(config.before)
+    const hookWrapper = adapter.wrapHook.bind(null, config, specs)
+
+    await hookWrapper(config.before)
     let result = await new Promise((resolve, reject) => {
         runner = mocha.run(resolve)
-        Object.keys(EVENTS).forEach((e) => runner.on(e, adapter.emit.bind(null, EVENTS[e], cid, capabilities)))
+        Object.keys(EVENTS).forEach((e) => runner.on(e, adapter.emit.bind(null, EVENTS[e], runner, cid, capabilities)))
+        runner.suite.beforeAll(hookWrapper(config.beforeSuite, runner, 'suite'))
+        runner.suite.beforeEach(hookWrapper(config.beforeTest, runner, 'test'))
+        runner.suite.afterEach(hookWrapper(config.afterTest, runner, 'test'))
+        runner.suite.afterAll(hookWrapper(config.afterSuite, runner, 'suite'))
     })
-    await runHook(config.after)
+    await hookWrapper(config.after)
     return result
 }
 
+adapter.wrapHook = function (config, specs, hookFn, runner, type) {
+    // Hooks with an instance of Mocha runner can access current test information, but
+    // need to call done() to notify async completion. Hooks outside Mocha context (e.g.
+    // before/after) are simpler.
+    if (runner) {
+        return function (done) {
+            let message = adapter.formatMessage({
+                payload: (type === 'test' ? runner.test : runner.suite.suites[0]),
```

Comment:
@christian-bromann I have no idea whether or not we can guarantee a single suite here?